[SBOM] Generate SBOMs for running containers

lebauce opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

SBOMs are already generated in the security module of system-probe. This PR introduces
a remote collector to serve and fetch SBOMs over gRPC, similar to the remote process collector.
SBOMs are associated to the workloadmeta `container` type. These SBOMs are then forwarded to
the intake through the SBOM check.

### Motivation

This allows vulnerabilities to be generated for software not part of the base image.

### Describe how you validated your changes

Enable runtime security in `system_probe.yaml`:

```
runtime_security_config:
  enabled: true

  sbom:
    enabled: true
```

Enable SBOM collection in `datadog.yaml`

```
sbom:
  enabled: true
```

- Restart the agent.
- Start a docker container with `docker run -ti ubuntu /bin/bash`
- Make sure the status of SBOM for the container is "SUCCESS" in `agent workload-list --verbose"

<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

lebauce committed:
Generate SBOMs for running containers from system-probe

lebauce committed:
Remove unused entity assignment

lebauce committed:
Index SBOMs based on container ID instead of container image

lebauce committed:
Create a dedicated gRPC service SBOM collection

lebauce committed:
Do not scan image if cache is not initialized

lebauce committed:
Extract cyclonedx conversion helpers to pkg/sbom/convert

lebauce committed:
Make linters happy (and me grumpy)

lebauce committed:
Serialize SBOM to JSON instead of protobuf

lebauce committed:
Fix linters

GustavoCaso commented on comp/core/workloadmeta/collectors/internal/remote/sbomcollector/sbom_collector.go:
```diff
@@ -0,0 +1,180 @@
+	}
+
+	if event.ID == "" {
+		return workloadmeta.Event{}, fmt.Errorf("expected container ID, got empty")
```
```suggestion
		return workloadmeta.Event{}, fmt.Errorf("expected event ID, got empty")
```

Maybe?

lebauce replied:
In this case, the event ID is really the container ID so I suggest to keep it as it as it would be easier to troubleshoot if that happens, does it ok to you ?

GustavoCaso replied:
Absolutely 😄

GustavoCaso commented on comp/core/workloadmeta/collectors/internal/remote/sbomcollector/sbom_collector.go:
```diff
@@ -0,0 +1,180 @@
+
+type client struct {
+	cl              sbompb.SBOMCollectorClient
+	parentCollector *streamHandler
```
It seems we are not using this field. Do we need it?

lebauce replied:
Indeed. Removed

GustavoCaso commented on comp/core/workloadmeta/impl/dump_test.go:
```diff
@@ -190,6 +194,8 @@ Health:
 Created At: 0001-01-01 00:00:00 +0000 UTC
 Started At: 0001-01-01 00:00:00 +0000 UTC
 Finished At: 0001-01-01 00:00:00 +0000 UTC
+----------- SBOM -----------
+SBOM is nil
```
Should we exercise a test in which the SBOM is not nil?

lebauce replied:
Added test

GustavoCaso commented on comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector.go:
```diff
@@ -145,6 +145,10 @@ func (s *streamHandler) Port() int {
 }
 
+func (s *streamHandler) Address() string {
+	return fmt.Sprintf(":%d", s.Port())
```
Should we make the `Port` function private? Since it is no longer part of the generic interface

lebauce replied:
:+1: 
Done

GustavoCaso commented on comp/core/workloadmeta/collectors/internal/remote/generic.go:
```diff
@@ -113,9 +113,17 @@ func (c *GenericCollector) Start(ctx context.Context, store workloadmeta.Compone
 		opts = append(opts, grpc.WithTransportCredentials(creds))
 	}
 
+	opts = append(opts, grpc.WithContextDialer(func(_ context.Context, url string) (net.Conn, error) {
+		if strings.HasPrefix(url, "/") {
+			return net.Dial("unix", url)
+		}
+
+		return net.Dial("tcp", url)
+	}))
```
I suggest making the generic implementation branch based on the URL provided via the address. I think we could probably create a new public function in the generic interface that each collector has to implement to return the correct connection struct

lebauce replied:
I can definitely add a `NewConnection` or `Connect` function to the generic interface.
Just to make sure I understand, why not modifying the generic implementation ? The socket vs tcp is an implementation detail.

GustavoCaso replied:
> Just to make sure I understand, why not modify the generic implementation? The socket vs tcp is an implementation detail.

The way I see it, the generic wrapper simply calls the function that the different implementations have. So instead of checking the return value of `Address()` we can abstract that and use something like:

```go
opts := []grpc.DialOption{grpc.WithContextDialer(func(_ context.Context, url string) (net.Conn, error) {
    return c.StreamHandler.NewConnection(url)
})}
```

The way the logic of which type of connection is pushed to each specific implementation.

Of course this is a possible way of doing it, and the way you've done it is valid as well 😄 But If I were making the change, I would go for `NewConnection`

GustavoCaso replied:
We talked over Slack and agree that @lebauce point are correct and we do not to duplicate the logic in the NewConnection 😄

GustavoCaso commented on pkg/security/resolvers/sbom/resolver.go:
```diff
@@ -366,53 +381,63 @@ func (r *Resolver) removePendingScan(containerID containerutils.ContainerID) {
+		RequestID: string(sb.ContainerID),
+		Report:    report,
 	}
+	r.NotifyListeners(SBOMComputed, scanResult)
```
Are we adding a test for this new functionality?

lebauce replied:
I added a test regarding the notification mechanism.
That being said, this functionality would deserve a bigger e2e test that I suggest to put in a dedicated PR.

GustavoCaso replied:
Thanks. Agree on a new e2e test on separate to avoid increasing the scope of this one 😄

lebauce committed:
Unexport Port method

lebauce committed:
Removed unused parentCollector field

lebauce committed:
Add unit test for non empty SBOM

lebauce committed:
Add functional test on notification mechanism

GustavoCaso commented on pkg/proto/datadog/sbom/workloadmeta_sbom.proto:
```diff
@@ -0,0 +1,20 @@
+    bytes Data = 1;
+    string Kind = 2;
+    // ID is the ID for this entity, in a format specific to the entity Kind.
+    string ID = 3;
```
Should we rename it to KindID to avoid confusion in the future?

lebauce replied:
Shamelessly stole it from https://github.com/DataDog/datadog-agent/blob/main/comp/core/workloadmeta/def/types.go#L227-L228

GustavoCaso commented on comp/core/workloadmeta/collectors/internal/remote/generic.go:
```diff
@@ -113,9 +113,17 @@ func (c *GenericCollector) Start(ctx context.Context, store workloadmeta.Compone
 
+	opts = append(opts, grpc.WithContextDialer(func(_ context.Context, url string) (net.Conn, error) {
+		if strings.HasPrefix(url, "/") {
+			return net.Dial("unix", url)
```
I was mostly asking because I saw this in the codebase 

https://github.com/DataDog/datadog-agent/blob/6cd58aa800614fba17cb7c455b84aac36843717f/pkg/security/agent/client.go#L184-L197

lebauce replied:
Based on offline discussions: UNIX sockets are supported on newer Windows so keeping it as it but using filepath.IsAbs instead

lebauce committed:
Change address check from HasPrefix from to IsAbs to support Windows

lebauce committed:
Fix usage of t.Errorf

lebauce committed:
Remove unused generic

lebauce commented on comp/core/workloadmeta/collectors/internal/remote/sbomcollector/sbom_collector.go:
```diff
@@ -0,0 +1,179 @@
+	return collectorEvents, nil
+}
+
+func handleEvents[T any](collectorEvents []workloadmeta.CollectorEvent, sbomEvents []T, convertFunc func(T) (workloadmeta.Event, error)) []workloadmeta.CollectorEvent {
```
No, because I was just lazy and copy/pasted the code from the existing remote collector.
Removed the generic

GustavoCaso reviewed (approved):
Awesome work

lebauce committed:
Only send components that are not parts of the base image

lebauce committed:
Add telemetry on components number

lebauce committed:
Fetch image from workloadmeta before pruning

lebauce committed:
Add logs

lebauce committed:
Fix wrong cast for ebpfless

lebauce committed:
Search image based on its name or repo digests

As stated in the comments for container image:
For historical reason, The imageId from containerd runtime and kubernetes refer to different fields.
For containerd, it is the digest of the image config.
For kubernetes, it referres to repo digest of the image (at least before CRI-O v1.28)

So we look for the image whose name is either the container image ID
or it has it in its repo digests

lebauce committed:
Fail when no OS component was found

lebauce committed:
Include container SBOMS in flare

lebauce committed:
Factorize SBOM output

lebauce committed:
Fetch image metadata only when needed

lebauce committed:
Forget request when we failed to fetch image metadata

lebauce committed:
Rebased on main

lebauce committed:
Do not use TLS for remote SBOM collector
